src/shared/hooks/use-linked-interactives-authoring.tsx

```diff
@@ -0,0 +1,106 @@
+  linkedInteractiveProps?: ILinkedInteractiveProp[];
+  schema: JSONSchema6;
+}
+
+type AuthoredState = {[key: string]: any};
+
+const emptyArray: ILinkedInteractive[] = [];
+
+export const useLinkedInteractivesAuthoring = ({ linkedInteractiveProps, schema }: IProps) => {
+  const initMessage = useInitMessage<AuthoredState>();
+  const [ interactiveList, setInteractiveList ] = useState<{[label: string]: {names: string[], ids: string[]}}>({});
+  const [ cachedLinkedInteractives, setCachedLinkedInteractives ] = useState<ILinkedInteractive[]>();
+  const { authoredState } = useAuthoredState<AuthoredState>();
+
+  // Note that initMessage.linkedInteractives is never updated, even after setLinkedInteractives is sent.
+  // So, cachedLinkedInteractives is used to keep the most recent value after updates.
+  const currentLinkedInteractives = cachedLinkedInteractives || (initMessage?.mode === "authoring" && initMessage?.linkedInteractives) || emptyArray;
+
+  // Handle authored state update. Each time one of the linked interactive properties is updated, send updated
+  // linkedInteractives array to LARA parent window. cachedLinkedInteractives acts only as a cache, so we don't
+  // send the message each time the authoredState is updated (e.g. its other, unrelated fields).
+  useEffect(() => {
+    if (linkedInteractiveProps && authoredState && initMessage?.mode === "authoring") {
+      linkedInteractiveProps.forEach(li => {
+        const name = li.label;
+        const authoredStateVal = authoredState[name];
+        const linkedInteractive = currentLinkedInteractives.find(l => l.label === name);
+        if (!linkedInteractive && authoredStateVal !== undefined) {
+          // Add a new item.
+          const newArray = currentLinkedInteractives.concat({
+            id: authoredStateVal,
+            label: name
+          });
+          setCachedLinkedInteractives(newArray); // Set cached value
+          setLinkedInteractives({linkedInteractives: newArray}); // Send to LARA
+        } else if (linkedInteractive && authoredStateVal === undefined) {
+          // Remove item from the array.
+          const idx = currentLinkedInteractives.indexOf(linkedInteractive);
+          const newArray = currentLinkedInteractives.slice();
+          newArray.splice(idx, 1);
+          setCachedLinkedInteractives(newArray); // Set cached value
+          setLinkedInteractives({linkedInteractives: newArray}); // Send to LARA
+        } else if (linkedInteractive && linkedInteractive.id !== authoredStateVal) {
+          // Update array item.
+          const idx = currentLinkedInteractives.indexOf(linkedInteractive);
+          const newArray = currentLinkedInteractives.slice();
+          newArray.splice(idx, 1, {
+            id: authoredStateVal,
+            label: name
+          });
+          setCachedLinkedInteractives(newArray); // Set cached value
+          setLinkedInteractives({linkedInteractives: newArray}); // Send to LARA
```

Comment:
There's a bug here, these calls should be outside the loop, and only one array should be generated. It won't be visible in practice, as an author could change only one field at the time, but still.

Reply:
Perhaps add a note in the code in case the situation comes up in the future.

Reply:
I've fixed that problem in the recent commits.